### PR TITLE
Rework of documentation

### LINKED ISSUE #17: Documentation rework
We need a better way of specifying and showing examples, rails integration, etc. instead of bloating the `README.md` file.

### PR DESCRIPTION
**Pull request by andreicek**
The documentation needed a bit more work. I've separated the rails part from the main `README.md` file and added the options and rails to documentation folder. Also the example documentation was added as well.

If merged this will fix #17.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,7 +5,7 @@ A missing link to Webpack and Ruby on Rails integration.
 
 [![Build Status](https://semaphoreci.com/api/v1/infinum/webpack-rails-manifest-plugin/branches/master/shields_badge.svg)](https://semaphoreci.com/infinum/webpack-rails-manifest-plugin) [![npm version](https://badge.fury.io/js/webpack-rails-manifest-plugin.svg)](https://badge.fury.io/js/webpack-rails-manifest-plugin)
 
-This plugin can be used to flush a list of your assets to a `manifest.json` file and replace asset pipeline. Using that file you can require your assets in Rails (see [Rails helper](#rails)).
+This plugin can be used to flush a list of your assets to a `manifest.json` file and replace asset pipeline. Using that file you can require your assets in Rails (see [Rails helper](documentation/rails.md)).
```
Comment (the-overengineer) [inline README.md]: replace **the** asset pipeline.

Using that file**,**

File: README.md [hunk 2]
```diff
@@ -27,29 +27,7 @@ const RailsManifestPlugin = require('webpack-rails-manifest-plugin');
...
-
-### mapAssetPath
-
-The function will receive three arguments:
-* `requirePath` (e.g. `images/photos/sunset.jpg`) - the path that was originally required in JS/CSS/HTML
-* `assetName` (e.g. `sunset.jpg`) - the file name that was originally required in JS/CSS/HTML
-* `isChunk` (eg. `true`) - specifies if the file is in the named chunk. Output JS and CSS files usually are.
-
-The function should return a string that will be used as a key in the manifest. By default, this will be the `requirePath` value.
+You'll find the `manifest.json` file in your output directory. More about options can be read [here](documentation/options.md).
```
Comment (the-overengineer) [inline README.md]: You can read more about the options here

File: README.md [hunk 3]
```diff
@@ -69,130 +47,20 @@ And then in `application.js` (your entrypoint)
...
-```
-
-This file would then be saved in `app/helpers/webpack_helper.rb` for example.
-
-Now you can use it like this:
-
-```HTML
-<img src="#{webpack_asset_url('logo.svg')}" alt="logo" />
-```
+Once you set everything up you should see this in your `manifest.json` file:
```
Comment (the-overengineer) [inline README.md]: everything up**,**

File: documentation/options.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
+Plugin options
+===============
+
+You can specify a few options to the plugin:
```
Comment (the-overengineer) [inline documentation/options.md]: Wording? Maybe: "The plugin allows you to specify several options:"

File: documentation/options.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+```JavaScript
+new RailsManifestPlugin({
```
Comment (the-overengineer) [inline documentation/options.md:7]: Note that the default values are given?

File: documentation/options.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  fileName: 'manifest.json',
+  writeToFileEmit: false,
+  extraneous: null,
+  mapAssetPath: (requirePath) => requirePath
+});
+```
+
+### fileName
+
+This is the output file that will be created after each change. The default is `manifest.json`.
```
Comment (the-overengineer) [inline documentation/options.md]: The default **value**, maybe?

File: documentation/options.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+### writeToFileEmit
+
+When using this plugin with [webpack-dev-plugin](#) manifest will not be written out to the filesystem, but to in-memory filesystem of the webpack-dev-plugin.
```
Comment (the-overengineer) [inline documentation/options.md]: Suggesting these changes:
webpack-dev-plugin**, the** manifest **file** will not be written **to a file**, but to **an** in-memory filesystem **used by** webpack-dev-plugin.

File: documentation/options.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+### extraneous
+
+If for any reason you wish to supply the files that will not go thru the Webpack comilation you can do so here. It's just important to note that they **will not** be copied over to the output folder.
```
Comment (the-overengineer) [inline documentation/options.md]: thru -> through
comilation -> compilation
Remove the **just** from just important

File: documentation/options.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  // ...
+  extraneous: ['./file.jpg'],
+  // ...
+}
+```
+
+### mapAssetPath
+
+The function will receive three arguments:
+* `requirePath` (e.g. `images/photos/sunset.jpg`) - the path that was originally required in JS/CSS/HTML
```
Comment (the-overengineer) [inline documentation/options.md:38]: Why the **originally**?

File: documentation/rails.md [hunk 10]
```diff
@@ -0,0 +1,124 @@
+Rails configuration
+===================
+
+To use this file with Rails you have two choices:
```
Comment (the-overengineer) [inline documentation/rails.md]: Reword. Maybe something like:
"Rails can be configured to use the manifest file in one of two ways"

File: documentation/rails.md [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+#### 1. Adding a new `webpack_asset_url` helper
+
+You'll need a helper to replace asset pipeline. Here is an example how you an do this:
```
Comment (the-overengineer) [inline documentation/rails.md]: The second sentence is nonsensical.

File: documentation/rails.md [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+  def manifest
+    @manifest ||= JSON.parse(File.read('manifest.json'))
+  rescue
+    fail 'Please run webpack'
+  end
+end
+```
+
+This file would then be saved in `app/helpers/webpack_helper.rb` for example.
```
Comment (the-overengineer) [inline documentation/rails.md]: Reword. Maybe:

"This module can then be saved in an appropriately named file (e.g. `app/helpers/webpack_helper.rb`)."

File: documentation/rails.md [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+
+Now you can use it like this:
```
Comment (the-overengineer) [inline documentation/rails.md:26]: You can now

Comment (the-overengineer) [reply documentation/rails.md:26]: Or rather "then"

File: documentation/rails.md [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+
+```HTML
+<img src="#{webpack_asset_url('logo.svg')}" alt="logo" />
+```
+
```
Comment (the-overengineer) [inline documentation/rails.md:31]: example **of**

File: documentation/rails.md [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+to require your assets. [Here](https://github.com/infinum/webpack-rails-manifest-plugin/blob/master/example/webpack.config.js#L12) is an example on how to add a digest to a file.
+
+**Take note that you can't use Rails `image_tag`, `stylesheet_link_tag`, `javascript_include_tag` helper.**
```
Comment (the-overengineer) [inline documentation/rails.md]: Take note -> Please note
**the** Rails [... helpers ...] helper**s**.

Also, maybe elaborate on the **can't** use? Do you mean that these helpers do not integrate with this plugin and will not function as expected?

File: documentation/rails.md [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+#### 2. Monkey patching rails assets helpers
```
Comment (the-overengineer) [inline documentation/rails.md]: **R**ails

File: documentation/rails.md [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+
+Be sure to remove rails sprockets entirely by editing your `applicaton.rb` file
```
Comment (the-overengineer) [inline documentation/rails.md:38]: **R**ails

File: documentation/rails.md [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+require "active_job/railtie"
+require "active_record/railtie"
+require "action_controller/railtie"
+require "action_mailer/railtie"
+require "action_view/railtie"
+# require "sprockets/railtie"
+require "rails/test_unit/railtie"
+```
+
+And after that copy the code below to your initializers folder `app/config/initializers/webpack-rails-manifest.rb`
```
Comment (the-overengineer) [inline documentation/rails.md]: Is that And in the beginning really necessary? Maybe just rephrase the whole thing a bit, this is also lacking punctuation. Maybe:

"Then save the following code in the initializers directory as `app/config/initializers/webpack-rails-manifest.rb`."

File: documentation/rails.md [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+          @manifest ||= JSON.parse(File.read('manifest.json'))
+        rescue
+          fail 'Please run webpack'
+        end
+    end
+  end
+end
+```
+
+Now, you can keep using rails helpers to get the assets
```
Comment (the-overengineer) [inline documentation/rails.md]: Suggesting rephrase:

"You can now use Rails helpers to get the assets, as you normally would."

File: documentation/rails.md [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+
+```Ruby
+= image_tag('logo.svg')
+
+= stylesheet_link_tag 'application', media: 'all'
+= javascript_include_tag 'application'
+```
+
+NOTE: this is an example for rails 4.2.6, the code above may vary on different rails versions
```
Comment (the-overengineer) [inline documentation/rails.md:124]: this example **uses** Rails 4.2.6 (although lowercase is also okay, as it's a gem name here)

File: example/README.md [hunk 21]
```diff
@@ -0,0 +1,51 @@
+Example Webpack setup
+======================
+
+Here you'll find an example Webpack configuration. The `webpack.config.js` here will output two bundles here: `application.js` and `vendor.js`, and will process an image.
```
Comment (the-overengineer) [inline example/README.md]: Maybe rephrase the whole thing a bit? E.g.:

Here we give an example Webpack configuration. The given `webpack.config.js` configuration file will output two bundles (`application.js` and `vendor.js`), as well as process images.

File: example/README.md [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+
+## Starting
+
+Clone the plugin and run `npm install` in the root folder and in the `example/` folder. Then start `npm run` to start Webpack. In the folder `build/` you should see four files:
```
Comment (the-overengineer) [inline example/README.md]: folder**,** and in the `example/`...
"start `npm run` to start" - Not wrong, but maybe not use "start" twice?
"You should now see four files in the `build/` directory." (or folder, whatever)

File: example/README.md [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+
+1. `application.js`,
+2. `vendor.js`,
+3. `963eb32907744d9a0d6b98127162808f.svg` (fingerprint will be different),
```
Comment (the-overengineer) [inline example/README.md]: Maybe use "file name" instead of "fingerprint"?

File: example/README.md [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+4. and `manifest.json` file.
+
+## Basic configuration
+
+The configuration is consited out of 4 main parts:
```
Comment (the-overengineer) [inline example/README.md]: **consists of four parts**

File: example/README.md [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+
+### 1. `entry`
+
+Here you define where are the JavaScript files where you will require all other JS code, SCSS (not in this example), and other assets. They are named here `application` and `vendor`. That are the names that they will take when outputed.
```
Comment (the-overengineer) [inline example/README.md]: Please rewrite this paragraph (or tell me to do it). While the gist of it is clear, the wording is pretty bad and not very grammatical.

File: example/README.md [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+
+### 2. `output`
+
+This is where you define the output path for all the Webpack generated files. Here your JS, SCSS, and other asset files will end up after the Webpack finishes.
```
Comment (the-overengineer) [inline example/README.md]: Webpack-generated

"Webpack will store the resulting JS, SCSS, and other asset files in the given directory after processing them" (or something along that vein)

File: example/README.md [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+
+### 3. `plugins`
+
+You define the `RailsManifestPlugin` with any other needed options. For all available options take a look at the [options documentation](../documentation/options.md).
```
Comment (the-overengineer) [inline example/README.md]: Maybe drop the **needed**

File: example/README.md [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+
+### 4. `module`
+
+Here are all the loaders used for your files. In this example only a `file-loader` is added for `.svg` files.
```
Comment (the-overengineer) [inline example/README.md]: Reword a bit? Maybe:

"The loaders used for loading and processing your files are defined here. In this example, only a..."

File: example/README.md [hunk 29]
```diff
@@ -0,0 +1,51 @@
...
+
+For a more complete picture of Webpack you should really take your time and read thru the [documentation](https://webpack.github.io/docs/).
```
Comment (the-overengineer) [inline example/README.md]: **_through**_

Also maybe documentation -> official documentation

File: example/README.md [hunk 30]
```diff
@@ -0,0 +1,51 @@
...
+
+## Output
+
+And you should get something like this out:
```
Comment (the-overengineer) [inline example/README.md]: Kill the **And** and **out**

File: example/README.md [hunk 31]
```diff
@@ -0,0 +1,51 @@
...
+{
+  "./logo.svg": "b69d5b43fc856f23ef7d8f0c8cc6b313.svg",
+  "application.js": "application.js",
+  "vendor.js": "vendor.js"
+}
+```
+
+## Integration with Rails
+
+For more info on how to completely replace asset pipeline read the [Rails integration documentation](../documentation/rails.md).
```
Comment (the-overengineer) [inline example/README.md]: info**rmation**
**the** asset pipeline

**Review by the-overengineer — CHANGES_REQUESTED** [review]

File: documentation/rails.md (hunk 11 above)
Comment (andreicek) [reply documentation/rails.md]: Yeah, it's a missed letter. 

> Here is an example how you can do this.

File: documentation/rails.md (hunk 15 above)
Comment (andreicek) [reply documentation/rails.md]: Paging @gabskoro here.

File: documentation/rails.md (hunk 20 above)
Comment (andreicek) [reply documentation/rails.md:124]: I don't agree with this one. It's not that example uses Rails, it's the other way around. We are changing the internals of Rails.

File: example/README.md (hunk 21 above)
Comment (andreicek) [reply example/README.md]: I've changed the first sentence to:

> Here is an example of a Webpack configuration.

Other - :+1:

File: example/README.md (hunk 22 above)
Comment (andreicek) [reply example/README.md]: Yup. Also it's not `npm run` it's `npm start`. Removed the first start.

:+1:

File: example/README.md (hunk 23 above)
Comment (andreicek) [reply example/README.md]: I'm not sure. I wan't to explain why there will be a bunch of random characters instead of a file name.

File: example/README.md (hunk 25 above)
Comment (andreicek) [reply example/README.md]: How about? 

> This is an array, or an object where you can list all the JavaScript files where you require every other asset (JS, SCSS, images, fonts...). Here we are listing two file, `application.js` and `vendor.js`, that are going to produce two outputs.

File: documentation/options.md (hunk 9 above)
Comment (andreicek) [reply documentation/options.md:38]: This was changed. Not sure why Github isn't showing this as outdated.

File: documentation/rails.md (hunk 14 above)
Comment (andreicek) [reply documentation/rails.md:31]: I'm not sure what you wanted to say here 😄.

Comment (the-overengineer) [reply documentation/rails.md:31]: My bad, one line down :smile:

File: documentation/rails.md (hunk 20 above)
Comment (the-overengineer) [reply documentation/rails.md:124]: Fair enough. How about a simple "We use rails 4.2.6 in this example", then?

File: example/README.md (hunk 23 above)
Comment (the-overengineer) [reply example/README.md]: Sure, but "fingerprint" means nothing if you don't know what this is about. How about _both_ - "The file name (also known as _fingerprint_) will be different)

File: example/README.md (hunk 25 above)
Comment (the-overengineer) [reply example/README.md]: Technically not true. You do not require everything else from them directly. But better, I guess. Won't be too anal about things that belong into the Webpack docs anyway.

File: documentation/rails.md (hunk 11 above)
Comment (the-overengineer) [reply documentation/rails.md]: Then at least "an example **of**"

File: example/README.md (hunk 23 above)
Comment (andreicek) [reply example/README.md]: Changed it to a:

> fingerprinted file name will be different

**Review by the-overengineer — APPROVED** [review]

File: README.md [hunk 32]
```diff
@@ -5,7 +5,11 @@ A missing link to Webpack and Ruby on Rails integration.
 
 [![Build Status](https://semaphoreci.com/api/v1/infinum/webpack-rails-manifest-plugin/branches/master/shields_badge.svg)](https://semaphoreci.com/infinum/webpack-rails-manifest-plugin) [![npm version](https://badge.fury.io/js/webpack-rails-manifest-plugin.svg)](https://badge.fury.io/js/webpack-rails-manifest-plugin)
 
-This plugin can be used to flush a list of your assets to a `manifest.json` file and replace asset pipeline. Using that file you can require your assets in Rails (see [Rails helper](#rails)).
+This plugin can be used to flush a list of your assets to a `manifest.json` file and replace the asset pipeline.
+
+## Rails
+
+Requireing assets in Rails will be a bit different and requires some configuration. Read thru our [Rails helper documentation](documentation/rails.md).
```
Comment (DarkoKukovec) [inline README.md]: through

File: documentation/options.md [hunk 33]
```diff
@@ -0,0 +1,42 @@
...
+});
+```
+
+### fileName
+
+This is the output file that will be created after each change. The default value is `manifest.json`.
+
+### writeToFileEmit
+
+When using this plugin with [webpack-dev-plugin](https://github.com/webpack/webpack-dev-server) the manifest file will not be written to a file, but to an in-memory filesystem used by webpack-dev-plugin.
```
Comment (DarkoKukovec) [inline documentation/options.md]: This statement doesn't say what it means if the property is set to true or false. Although it can be figured out from the property name, it's not 100% clear.

File: example/README.md [hunk 34]
```diff
@@ -0,0 +1,51 @@
+Example Webpack setup
+======================
+
+Here is an example of a Webpack configuration. The given `webpack.config.js` configuratio nfile will output two bundles (`application.js` and `vendor.js`), as well as process an image.
```
Comment (DarkoKukovec) [inline example/README.md]: *configuration file

File: README.md (hunk 32 above)
Comment (the-overengineer) [reply README.md]: Also, "Requiring". And while through is much better than thru, I'm not sure it's needed here at all. Works just the same if you kick the word out.

Comment (andreicek) [reply README.md]: Removed the word thru, and changed the requires to needs.

Comment (the-overengineer) [reply README.md]: I meant "Requireing" -> "Requiring", if that was unclear.
